biocrnpyler/species.py

```diff
@@ -878,19 +796,461 @@ def replace(self, position, part, direction=None):
+        3.  A single Species S has a parent which is an OrderedPolymerSpecies with a PolymerConformation as a parent.
+            Produces: a (Ordered)ComplexSpecies in a PolymerConformation with in the OrderedPolymerSpecies inside the PolymerConformation
+        4.  Multiple Species S have parents which are OrderedPolymerSpecies without parents.
+            Produces: a (Ordered)ComplexSpecies containing all S inside a PolymerConformation
+        5.  One or More Species S have parents which are OrderedPolymerSpecies with parents and/or PolymerConformations
+            Produces: a (Ordered)ComplexSpecies containing all S inside a PolymerConformation which merges all PolymerComformation Complexes.
+        """
+        species = []
+        # below is extracting the "species" keywords from the args
+        keywarg = None
+        if("species" in keywords):
+            keywarg = True
+            species = keywords["species"]
+            del keywords["species"]
+        elif(len(args) >= 1):
+            keywarg = False
+            species = args[0]
+            args = args[1:]
+
+        if not isinstance(species, list):
+            raise TypeError(f"First argument to Complex (or species keyword), must be a list of Species; recieved {species}.")
+
+        # Check whether ot make a ComplexSpecies or OrderedComplexSpecies
+        if "ordered" in keywords:
+            ordered = keywords["ordered"]
+            del keywords["ordered"]
+            ComplexClass = OrderedComplexSpecies
+        else:
+            ordered = False
+            ComplexClass = ComplexSpecies
+
+        #Use to supress errors in ComplexSpecies and OrderedComplexSpecies
+        keywords["called_from_complex"] = True
+
+        parent_species = []  # parent_species is a list of OrderedPolymerSpecies and/or PolymerConformations
+        bindlocs = []  # bindloc is the location a Species is bound to parent_species
+        insertlocs = [] #insertloc is the location of the species inside the Complex.species list (important to maintain order in OrderedComplexSpecies)
+        child_species = [] #the species with parents
+        other_species = []  # Other species in the Complex
+
+        # Below cycle through species and see if one has a parent. If it does, that means the species is
+        # in an OrderedPolymerSpecies and the Complex should be formed around it.
+        for i, specie in enumerate(species):
+            if(hasattr(specie, "parent") and (specie.parent is not None)):
+                #This adds to a growing list of parents, which will be placed in an PolymerConformation
+                # It is very important to deepcopy here because the underlying OrderedPolymerSpecies or PolymerConformation will be modified.
+                parent_species.append(specie.parent)
+                bindlocs.append(specie.position)
+                insertlocs.append(i)
```

Comment:
insertlocs tells where in the list species the 'specie' with a parent came from. There may be multiple "specie"'s all with parents and I keep track of each ones index in the list species in order to preserve the order of the list later. This is something I added to prevent Compex from changing the order of lists (which could matters when making OrderedComplexSpecies).